### ISSUE TITLE
Feature: include controls for dependency management

### ISSUE THREAD

**Issue by zkamvar** [carpentries/sandpaper#21]
At the moment, we are thinking of using [{renv}] for dependency management, but there is the question of exactly how much we want to be responsible for in terms of supporting languages.

For example, if we are supporting an R lesson, we have the correct solution with {renv}, but if we wanted to support the environment for a python lesson, then it becomes tricker because of the relationship between {renv} and {reticulate} (See https://github.com/rstudio/renv/issues/543 and https://github.com/rstudio/renv/issues/537)

We know that BASH works and we are _fairly_ confident that SQL works, but the question is: how do we ensure that these lessons can have auto-generated output should the lesson maintainer choose to do so? 

[{renv}]: https://rstudio.github.io/renv

**Comment by fmichonneau** [comment]
We may want to split this issue into multiple:
- use {renv} for R-based lessons (target for beta release)
- have a solution for Python-based lessons (for v1)
- have a more general solution (for later)

could we detect and act on a `requirements.txt` or a `packages.json` file?

**Comment by zkamvar** [comment]
> We may want to split this issue into multiple:
> 
>     * use {renv} for R-based lessons (target for beta release)
> 
>     * have a solution for Python-based lessons (for v1)
> 
>     * have a more general solution (for later)
> 
> 
> could we detect and act on a `requirements.txt` or a `packages.json` file?

I think this is a reasonable approach and in fact, we could take a page out of Binder's playbook: https://mybinder.readthedocs.io/en/latest/using/config_files.html

The only thing that we need to think about is how these lockfiles will be updated. For R, it's a matter of deleting the lockfile and generating a new one with current versions of the packages, but for Python, where packages are more micromanaged with their versions, I don't know if it's that simple.

**Comment by zkamvar** [comment]
Some research:

{renv} was always kind of intimidating to me because it would embed an `activate.R` and `.Rprofile` from the start that would force the user to only use the packages in the lockfile and would error if you tried to use a package that wasn't in the lockfile. It makes sense, but can be very frustrating. 

Right at the start of `rstudio::conf(2021)` (luckily _right_ before I had to call in sick), I saw this tweet: https://twitter.com/grrrck/status/1352295743701340160. I knew I needed to get back to it, and I think it may be worthwhile:

When you dig down into the comments, the poster points to this repository: https://github.com/JosiahParry/renv-example#readme, which demonstrates a situation where {renv} can be used as JUST a lockfile (I'm testing this out at the moment..... and I should have read before I ran the code.... if there is no `renv/` directory, it will update packages to your GLOBAL library :grimace:, and you can use `renv::init()` to keep things local).

Moreover, [Miles McBain points out that {renv} is really flexible about how you can implement it](https://twitter.com/MilesMcBain/status/1352608733671550977). 

SO, it seems that we might be able to provide a way for continously updating {sandpaper}, {varnish}, and {pegboard} on GitHub while locking down these dependencies.

**Comment by zkamvar** [comment]
One more resource that is key: RStudio has created a guide for reproducing environments and it's a good exercise to figure out where we fall on this chart. https://environments.rstudio.com/reproduce.html

**Comment by zkamvar** [comment]
I JUST LOST A LONG WRITEUP OF MY THOUGHTS AFTER PLAYING WITH RENV THAT I WILL NEVER GET IT BACK :weary:

**Comment by zkamvar** [comment]
I think I might be able to summarize:

1. We can have a single `renv.lock` in our build stack and then {renv} will help bootstrap that for users (but we should provide the interface)
2. Updating the lockfile is as simple as `renv::update()`, which ties in nicely with #36 (Caveat: the github API is hit if we include github packages in there)
3. Bootstrapping/updating our own tooling is not immediately straightforward since {renv} likes to default to a closed system. 
4. Users need to be able to work offline (that means no ["It just started updating and I couldn't stop it"](https://users.umiacs.umd.edu/~tdumitra/courses/ENEE657/Fall17/papers/Vaniea16.pdf) situations). See https://github.com/carpentries/styles/pull/495#issuecomment-714568811 for Maxim's thoughts on this.
4. Whatever we decide, we need to make sure we stay up-to-date on our deployment system and allow users to choose when they want to update our build system.

**Comment by zkamvar** [comment]
Oh hey! Good things come to those who wait.

{renv} just got a shiny new update, which includes the concept of different renv profiles: https://rstudio.github.io/renv/articles/profiles.html. How exactly this is going to work, I'm not sure, but I am excite!

I also found an issue that describes the concept I'm thinking about in this workflow: https://github.com/rstudio/renv/issues/624

**Comment by zkamvar** [comment]
I was wondering where these profiles would live and I have my answer:

1. use `renv::snapshot()` to create a `renv.lock` of the project
2. use `renv::activate(profile = "sandpaper")`, which will create the `renv/` folder with `renv/profiles/sandpaper/` 
3. use `renv::snapshot()` again to create a `renv.lock` inside of `renv/profiles/sandpaper/`

With this setup, we can create a few profiles:

default: stable profile that contains sandpaper and all the packages needed to build the lesson. This is the one that can be updated via #36 
development: profile that maintainers can use when they want to check out package changes. Ideally they should be able to revert.
sandpaper: This is an ephemeral lockfile that is a copy of the default lockfile, but allows us to update sandpaper and its dependencies with #35 periodically. Users should be able to fix build versions in this profile if they want to.

**Comment by zkamvar** [comment]
After some playing around, I realized that using {renv} in conjunction with {callr} makes for a consistent environment without needing to subject the user to the confines of {renv} (and may even bring them to a better solution if things begin to break). 

Proof:

```r
# Download a renv.lock file into a temporary directory
tmp <- tempfile() 
dir.create(tmp) 
download.file("https://raw.githubusercontent.com/JosiahParry/renv-example/master/renv.lock", file.path(tmp, "renv.lock")) 

# Use callr to initialize a renv library from the global cache
callr::r(function(x) { 
    one <- try(packageVersion("sandpaper")) 
    renv::init(x) 
    two <- tryCatch(packageVersion("sandpaper"), error = function(e) {e}) 
    print(two) 
    message("installed packages: ", nrow(installed.packages())) 
    message("library length: ", length(dir(renv::paths$library()))) 
    if (inherits(one, "try-error")) two else one 
  }, args = list(x = tmp), show = TRUE)  
```

**Comment by zkamvar** [comment]
Update: This is now being addressed in #158
